### PR TITLE
Adding hamming distance function to painless for dense_vector fields

### PR DESCRIPTION
**Pull request by benwtrent**
This adds `hamming` distances, the pop-count of `xor` byte vectors as a first class citizen in painless. 

For byte vectors, this means that we can compute hamming distances via script_score (aka, brute-force).

The implementation of `hamming` is the same that is available in Lucene, and when lucene 9.11 is merged, we should update our logic where applicable to utilize it.

NOTE: this does not yet add hamming distance as a metric for indexed vectors. This will be a future PR after the Lucene 9.11 upgrade.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by pmpailis — APPROVED** [review]
LGTM

**Review by carlosdelest — APPROVED** [review]
👍

File: server/src/main/java/org/elasticsearch/script/field/vectors/BinaryDenseVector.java [hunk 1]
```diff
@@ -83,6 +83,16 @@ public double l1Norm(List<Number> queryVector) {
         return l1norm;
     }
 
+    @Override
+    public int hamming(byte[] queryVector) {
+        throw new UnsupportedOperationException("hamming distance is not supported for float vectors");
+    }
```
Comment (tteofili) [inline server/src/main/java/org/elasticsearch/script/field/vectors/BinaryDenseVector.java:89]: I was assuming that this being a `byte` query vector, it should not throw an `UOE` related to float vectors?

File: server/src/main/java/org/elasticsearch/script/field/vectors/KnnDenseVector.java [hunk 2]
```diff
@@ -85,6 +85,16 @@ public double l1Norm(List<Number> queryVector) {
         return result;
     }
 
+    @Override
+    public int hamming(byte[] queryVector) {
+        throw new UnsupportedOperationException("hamming distance is not supported for float vectors");
+    }
```
Comment (tteofili) [inline server/src/main/java/org/elasticsearch/script/field/vectors/KnnDenseVector.java:91]: same as above.

File: server/src/main/java/org/elasticsearch/script/field/vectors/BinaryDenseVector.java (hunk 1 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/script/field/vectors/BinaryDenseVector.java:89]: Its a `byte[]` query vector against a `float` indexed vector.

File: server/src/main/java/org/elasticsearch/script/field/vectors/KnnDenseVector.java (hunk 2 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/script/field/vectors/KnnDenseVector.java:91]: Its a `byte[]` query vector against a `float` indexed vector.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**